src/Components/Components/src/NavigationManagerExtensions.cs

```diff
@@ -0,0 +1,563 @@
+                {
+                    parameterData.DidReplace = true;
+
+                    if (parameterData.EncodedValue is not null)
+                    {
+                        newQueryStringBuilder.AppendParameter(pair.EncodedName.Span, parameterData.EncodedValue);
+                    }
+                }
+                else
+                {
+                    newQueryStringBuilder.AppendParameter(pair.EncodedName.Span, pair.EncodedValue.Span);
+                }
+            }
+
+            // Append any parameters with non-null values that did not replace existing parameters.
+            foreach (var (encodedName, data) in parameterDataByEncodedName)
+            {
+                if (!data.DidReplace && data.EncodedValue is not null)
+                {
+                    newQueryStringBuilder.AppendParameter(encodedName.Span, data.EncodedValue);
+                }
+            }
+
+            return newQueryStringBuilder.UriWithQueryString;
+        }
+
+        private static string UriWithAppendedQueryParameters(
+            string uriWithoutQueryString,
+            IReadOnlyDictionary<string, object?> parameters)
+        {
+            var builder = new QueryStringBuilder(uriWithoutQueryString);
+
+            // Build a new query from the existing URI, appending all parameters with non-null values.
+            foreach (var (name, value) in parameters)
+            {
+                var encodedName = System.Uri.EscapeDataString(name);
+                var encodedValue = GetEncodedParameterValue(value);
+
+                if (encodedValue is not null)
+                {
+                    builder.AppendParameter(encodedName, encodedValue);
+                }
+            }
+
+            return builder.UriWithQueryString;
+        }
+
+        private static string? GetEncodedParameterValue(object? value)
+        {
+            if (value is null)
+            {
+                return null;
+            }
+
+            var parameterType = value.GetType();
+
+            if (!_queryParameterFormatters.TryGetValue(parameterType, out var formatter))
+            {
+                throw new InvalidOperationException($"Cannot add query parameter of type '{parameterType}'.");
+            }
+
+            var formattedValue = formatter(value);
+            return formattedValue is null ? null : System.Uri.EscapeDataString(formattedValue);
+        }
+
+        private static bool TryRebuildExistingQueryFromUri(
+            string uri,
+            out QueryStringEnumerable existingQueryStringEnumerable,
+            out QueryStringBuilder newQueryStringBuilder)
+        {
+            var queryStartIndex = uri.IndexOf('?');
+
+            if (queryStartIndex < 0)
+            {
+                existingQueryStringEnumerable = default;
+                newQueryStringBuilder = default;
+                return false;
+            }
```

Comment:
Question: is it at all common to have URIs where there is no querystring, but there is still a trailing '?' (i.e. a query with no parameters)?

Currently, if something like that is encountered, we treat it as there being a full querystring and we don't take the shortcut route of not parsing the query part of the URI.

If something like this _is_ common, we could consider adding some logic to take the shortcut path in that case as well.

Reply:
Is optimizing for it as simple as treating `queryStartIndex == uri.Length - 1` the same as `queryStartIndex  < 0`? If so, sure, let's do it. If there's more involved, we can punt it until we find it compelling.

Reply:
It's not quite that simple, since the final URI would then have an extra `?`. We could change the `QueryStringBuilder` implementation to explicitly check for that, but it wouldn't handle the special case where we're adding a single parameter to an empty query with a `?` in it (since we skip using `QueryStringBuilder` altogether). We could special-case that logic as well, but at that point it's getting a bit messy. Another alternative is we pass a `ref Memory<char>` to `TryRebuildExistingQueryFromUri` as the URI rather than a `string`, and that method would slice the `?` out of the URI if it there are no query parameters. My concern there is that `TryRebuildExistingQueryFromUri` would IMO have too many responsibilities at that point.

So, it's possible, but I just haven't found a clean implementation for it yet.

Reply:
> So, it's possible, but I just haven't found a clean implementation for it yet.

Let's skip this scenario for now. It's only slower for a fairly edge case, and we can live with that.